cmd/run.go

```diff
@@ -756,3 +757,55 @@ func sendTelemetry(s *scheduler.Scheduler, c *cli.Context) {
 
 	telemetry.SendEventWithAssetStats("run_assets", assetStats, c)
 }
+
+func ExcludeAssetsByTag(excludeTag string, p *pipeline.Pipeline, s *scheduler.Scheduler, assetsByTag []*pipeline.Asset) int {
+	var excludedAssets []*pipeline.Asset
+	if len(assetsByTag) != 0 {
+		excludedAssets = p.GetAssetsByTagFromSubset(excludeTag, assetsByTag)
+	} else {
+		excludedAssets = p.GetAssetsByTag(excludeTag)
+	}
+	// Deduplicate excluded assets
+	uniqueAssets := make(map[string]*pipeline.Asset)
+	for _, asset := range excludedAssets {
+		uniqueAssets[asset.Name] = asset
+	}
+
+	if len(uniqueAssets) == 0 {
+		_, _ = errorPrinter.Printf("No assets found with the tag '%s' to exclude.\n", excludeTag)
+		return 0
+	}
+
+	for _, asset := range uniqueAssets {
+		s.MarkAsset(asset, scheduler.Succeeded, false) // Mark as succeeded to skip execution
+	}
+
+	_, _ = infoPrinter.Printf("Excluded %d assets with tag '%s'.\n", len(uniqueAssets), excludeTag)
+	return len(uniqueAssets)
+}
+
+func handleIncludeTag(tag string, p *pipeline.Pipeline, s *scheduler.Scheduler, runDownstreamTasks bool) ([]*pipeline.Asset, error) {
+	var assetsByTag []*pipeline.Asset
+
+	if tag != "" {
+		assetsByTag = p.GetAssetsByTag(tag)
+
+		if len(assetsByTag) == 0 {
+			errorPrinter.Printf("No assets found with the tag '%s'\n", tag)
+			return nil, fmt.Errorf("no assets found with tag '%s'", tag)
+		}
+
+		// Use infoPrinter or another custom printer for debug information
+		infoPrinter.Printf("Marking assets with tag '%s' to run.\n", tag)
+
+		// Mark assets with the given tag
+		s.MarkAll(scheduler.Succeeded)
+		s.MarkByTag(tag, scheduler.Pending, runDownstreamTasks)
+
+		infoPrinter.Printf("Running only the assets with tag '%s', found %d assets.\n", tag, len(assetsByTag))
+	} else {
+		assetsByTag = []*pipeline.Asset{}
+	}
+
+	return assetsByTag, nil
```

Comment:
I don't like this idea that this function is returning an irrelevant value that is used by another function, this seems like an optimization that could be done elsewhere. could we encapsulate this logic into another struct that's solely responsible from filtering what'll run?